sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -169,11 +169,21 @@ final class DataFrameWriter private[sql](df: DataFrame) {
   private def insertInto(tableIdent: TableIdentifier): Unit = {
     val partitions = partitioningColumns.map(_.map(col => col -> (None: Option[String])).toMap)
     val overwrite = mode == SaveMode.Overwrite
+
+    // A partitioned relation schema's can be different from the input logicalPlan, since
+    // partition columns are all moved after data column. We Project to adjust the ordering.
+    // TODO: this belongs in the analyzer.
+    val input = partitioningColumns.map { parCols =>
+      val projectList = df.logicalPlan.output.filterNot(c => parCols.contains(c.name)) ++
+        parCols.map(UnresolvedAttribute(_))
+      Project(projectList, df.logicalPlan)
```

Comment:
I found this block is a little bit hard to understand. The following version may be questionably more readable:

```scala
      val (inputDataCols, inputPartCols) = df.logicalPlan.output.partition {
        c => partCols.contains(c.name)
      }
      val projectList = inputDataCols ++ inputPartCols.map(c => UnresolvedAttribute(c.name))
      Project(projectList, df.logicalPlan)
```